packages/aws-cdk/lib/api/garbage-collector.ts

```diff
@@ -0,0 +1,393 @@
+    return info.bucketName;
+  }
+
+  private async getBootstrapQualifier(sdk: ISDK, bootstrapStackName: string): Promise<string> {
+    const info = await ToolkitInfo.lookup(this.props.resolvedEnvironment, sdk, bootstrapStackName);
+    return info.bootstrapStack.parameters.Qualifier;
+  }
+
+  /**
+   * Generator function that reads objects from the S3 Bucket in batches.
+   */
+  private async *readBucketInBatches(s3: S3, bucket: string, batchSize: number = 1000): AsyncGenerator<S3Asset[]> {
+    let continuationToken: string | undefined;
+
+    do {
+      const batch: S3Asset[] = [];
+
+      while (batch.length < batchSize) {
+        const response = await s3.listObjectsV2({
+          Bucket: bucket,
+          ContinuationToken: continuationToken,
+          MaxKeys: batchSize - batch.length,
+        }).promise();
+
+        response.Contents?.forEach((obj) => {
+          if (obj.Key) {
+            batch.push(new S3Asset(bucket, obj.Key));
+          }
+        });
+
+        continuationToken = response.NextContinuationToken;
+
+        if (!continuationToken) break; // No more objects to fetch
+      }
+
+      if (batch.length > 0) {
+        yield batch;
+      }
+    } while (continuationToken);
+  }
+
+  /**
+   * Fetches all relevant stack templates from CloudFormation. It ignores the following stacks:
+   * - stacks in DELETE_COMPLETE or DELET_IN_PROGRES stage
+   * - stacks that are using a different bootstrap qualifier
+   *
+   * It fails on the following stacks because we cannot get the template and therefore have an imcomplete
+   * understanding of what assets are being used.
+   * - stacks in REVIEW_IN_PROGRESS stage
+   */
+  private async fetchAllStackTemplates(cfn: CloudFormation, qualifier: string) {
+    const stackNames: string[] = [];
+    await paginateSdkCall(async (nextToken) => {
+      const response = await cfn.listStacks({ NextToken: nextToken }).promise();
+
+      // We cannot operate on REVIEW_IN_PROGRESS stacks because we do not know what the template looks like in this case
+      const reviewInProgressStacks = response.StackSummaries?.filter(s => s.StackStatus === 'REVIEW_IN_PROGRESS') ?? [];
+      if (reviewInProgressStacks.length > 0) {
+        throw new Error(`Stacks in REVIEW_IN_PROGRESS state are not allowed: ${reviewInProgressStacks.map(s => s.StackName).join(', ')}`);
+      }
+
+      // Deleted stacks are ignored
+      stackNames.push(
+        ...(response.StackSummaries ?? [])
+          .filter(s => s.StackStatus !== 'DELETE_COMPLETE' && s.StackStatus !== 'DELETE_IN_PROGRESS')
+          .map(s => s.StackId ?? s.StackName),
+      );
+
+      return response.NextToken;
+    });
+
+    print(chalk.blue(`Parsing through ${stackNames.length} stacks`));
+
+    const templates: string[] = [];
+    for (const stack of stackNames) {
+      let summary;
+      summary = await cfn.getTemplateSummary({
```

Comment:
All CFN calls need to be protected against throttling. Their limit is very low, 2tps or what have you. Guaranteed we will easily hit those limits in real life scenarios.

Reply:
I _think_ we are already doing this by default in CFN. It looks like we are using this:

https://github.com/aws/aws-cdk/blob/73d8eda290d70715235c8f9dcd2e4001a736534c/packages/aws-cdk/lib/api/aws-auth/sdk.ts#L157-L162

and we've already put some thought into retry handling here: https://github.com/aws/aws-cdk/pull/8711/files